janitor/functions.py

```diff
@@ -3869,3 +3813,230 @@ def sort_naturally(
     """
     new_order = index_natsorted(df[column_name], **natsorted_kwargs)
     return df.iloc[new_order, :]
+
+
+@pf.register_dataframe_method
+def expand_grid(df: pd.DataFrame = None,
+                df_key: str = None,
+                others: Dict = None) -> pd.DataFrame:
+    """ Creates a dataframe from all combinations of all inputs.
+        This method mutates the original dataframe.
+
+    Functional usage syntax:
+
+    .. code-block:: python
+
+        import pandas as pd
+        import janitor as jn
+
+        df = pd.DataFrame(...)
+        df = jn.expand_grid(df=df, df_key="...", others={...})
+
+    Method-chaining usage syntax:
+
+        import pandas as pd
+        import janitor as jn
+
+    .. code-block:: python
+
+        df = pd.DataFrame(...).expand_grid(df_key="bla",others={...})
+
+    :param df: A pandas dataframe.
+    :param df_key: name of key for the dataframe. It becomes the column name of the dataframe.
+    :others: A dictionary. This will contain the data to be combined with the dataframe.
+    :returns : A pandas dataframe of all combinations of name value pairs.
+
+    This works with a dictionary of name value pairs, and will work with structures that are not dataframes.
+    The output will always be a dataframe.
+    """
+    #check if others is a dictionary
+    if not isinstance(others, dict):
+        #strictly name value pairs
+        #same idea as in R and tidyverse implementation
+        #probably take this out,
+        #as it is covered in expand_grid function
+        raise ValueError("others must be a dictionary")
+    #if there is a dataframe, for the method chaining,
+    #it must have a key, to create a name value pair
+    if df is not None:
+        if not df_key:
+            raise ValueError("dataframe requires a name")
+        else:
+            others.update({df_key: df})
+    dfs, dicts = check_instance(others)
+
+    return grid_computation(dfs, dicts)
+
+
+def check_instance(entry):
```

Comment:
We need a docstring for this function. 

Additionally, I think it possibly could be moved out of the `functions.py` file into a utilities file, as this is not directly a monkey-patched function on top of a dataframe.